### PR TITLE
Migrating from node_redis to ioredis.

### PR DESCRIPTION
**Pull request by kulbirsaini**
Related Issue: https://github.com/OptimalBits/bull/issues/185

I have started this migration from node_redis to ioredis. Please review and suggest changes.

**Pending tasks**
- [ ] Improve/fix tests
- [ ] Update Readme
- [ ] Bump version since breaking change

**Status**
All tests passing with a few timing out when run with `npm test`.

*Borrows changes from PR #158

### CODE REVIEW

**Comment by nodesocket** [comment]
:+1: thanks for re-trying this. I couldn't get any of the tests to pass, but you seem to have cracked it.

Look forward to having this merged, so we can upgrade.

**Comment by manast** [comment]
You did  great job!.
Any ideas about the failing tests?

**Comment by kulbirsaini** [comment]
@manast please don't merge it yet. I am working on improving tests. Will do that step by step.

**Comment by kulbirsaini** [comment]
@manast as of now I have some tests which fails when I run the entire suite but they pass individually. I notice that they fail under load. Will it be possible to for you to fetch this and run tests on your machine?

@albertjan can you please help here? :-)

@nodesocket I had encountered that error previously and knew what was wrong when I looked at your PR. You get that error when you `JSON.parse` fails.

**Comment by albertjan** [comment]
I don't really understand why we would want to switch to ioredis? node_redis does it's job doesn't it?

**Comment by kulbirsaini** [comment]
@albertjan please see https://github.com/OptimalBits/bull/issues/185

**Comment by albertjan** [comment]
The time spent in the `process` function is almost always going to be the place where the software spends most of its time, then after that we have bull which does logic, making the performance gains even less. That point is moot if you ask me. 

I don't know enough about redis clusters to have an opinion about it. Why would you want a cluster? You plan to have so many jobs at one time in the database that it won't fit into memory? Or because you want the availability of multiple redis-databases? We've been running bull with redis in production for almost a year now. And we've never had issues with redis being down. 

I'll look into why your tests are failing but we rely heavily on the stability of bull and introducing a new redis client is a big change.

**Comment by kulbirsaini** [comment]
@albertjan personally, I don't care much about redis's performance and cluster part. I don't have an app at a scale large enough to worry about any of those two. I personally use `ioredis` (because of Bluebird) in my apps (for usage other than Bull) and more and more people seems to be interested in ioredis. People have similar requests on `Kue` as well.

I thought it'd be a good Sunday project so just tried this last night. If it works, everyone wins otherwise it was just a Sunday :smile:

**Comment by albertjan** [comment]
Got it :smile: as I said I'll look at it too when I have some time.

**Comment by thelinuxlich** [comment]
+1

**Comment by nodesocket** [comment]
:+1:

**Comment by julianfresco** [comment]
Hi, has this effort been abandoned? Any updates since October 2015? I'm very interested in this switch to ioredis.

Basically, our sysadmins want us to switch over to ioredis so they can make use of Redis Sentinel for the Automatic Failover feature (which is a huge win for distributed systems) but that would clearly not work as expected with Bull on node_redis.

**Comment by manast** [comment]
I have checkef ioredis and could not see any major wins. Regarding sentinel, I have to check why it would not work on node redis.

**Comment by manast** [comment]
ok, It seems as node_redis only can support sentinel via wrappers and no imminent plan to support it is there in the roadmap. On the other hand ioredis may be less mature than node_redis, so the perfect solution would be to be able to support both, I will see what I can do.

**Comment by julianfresco** [comment]
Thanks for the quick update @manast

**Comment by wprater** [comment]
we need ioredis support for Sentinel.  Any traction on this?

**Comment by barczaG** [comment]
Hey hey, I'm just amazed the efficiency of bull, but to use it in production we would need cluster support, do you guys plan to merge this PR?

**Comment by xycloud** [comment]
waiting so long

**Comment by manast** [comment]
if bull worked properly with ioredis and all unit test passings I will not have any reason why not to merge it, but this is not yet the case unfortunately...

**Comment by doublerebel** [comment]
Is anyone actively working on this?  If so please mention it, otherwise I can fork this PR and get all tests passing this weekend.

File: lib/job.js [hunk 1]
```diff
@@ -35,10 +35,10 @@ Job.create = function(queue, jobId, data, opts){
 
 Job.fromId = function(queue, jobId){
   return queue.client.hgetallAsync(queue.toKey(jobId)).then(function(jobData){
-    if(jobData){
-      return Job.fromData(queue, +jobId, jobData);
+    if(_.isEmpty(jobData)){
```
Comment (bradvogel) [inline lib/job.js]: Remove this. It's just hiding an error (https://github.com/OptimalBits/bull/pull/275/files) and leaves corrupted data in Redis, causing the job processor to get stuck in a loop.

Comment (doublerebel) [inline lib/job.js]: Awesome, thanks for the tip.

**Comment by doublerebel** [comment]
Really close to having this completed, only 4 tests remain not passing.
https://travis-ci.org/nextorigin/bull/jobs/175246125

The majority of unhandledRejection warnings I've had to squash, are caused by Bull attempting a Redis operation after the connection is closed.  Promises unfortunately leave a vague stacktrace, so I'm going to take a break and look at this again tomorrow.

**Comment by manast** [comment]
will close since ioredis support has been merged to master

**Comment by xycloud** [comment]
awesome

**Pull request closed without merging**